File: kalitech-engine/src/main/resources/kalitech/engine/modules/Log/Log.js
// FILE: resources/kalitech/builtin/Log.js
// Author: Calista Verner
"use strict";

function safeJson(v) {
    try {
        return JSON.stringify(v);
    } catch (_) {
    }
    try {
        return String(v);
    } catch (_) {
    }
    return "[unserializable]";
}

function isThrowableLike(v) {
    if (!v) return false;

    try {
        if (v instanceof Error) return true;
    } catch (_) {
    }

    const t = typeof v;
    if (t !== "object" && t !== "function") return false;

    try {
        const stack = v.stack;
        if (typeof stack === "string" && stack.length > 0) return true;
    } catch (_) {
    }

    try {
        const name = v.name;
        const msg = v.message;
        if (typeof name === "string" && name.length > 0 && typeof msg === "string") return true;
    } catch (_) {
    }

    // Host exceptions (PolyglotException/Throwable) sometimes expose Java-ish surface
    try {
        if (typeof v.getClass === "function") return true;
    } catch (_) {
    }

    return false;
}

function throwableToText(e) {
    if (!e) return "null";

    try {
        if (e instanceof Error) return (e.stack || e.message || String(e));
    } catch (_) {
    }

    try {
        if (typeof e.stack === "string" && e.stack) return e.stack;
    } catch (_) {
    }

    try {
        if (typeof e.message === "string" && e.message) return e.message;
    } catch (_) {
    }

    try {
        return String(e);
    } catch (_) {
    }

    return "[unserializable-exception]";
}

function valueToText(v) {
    if (v == null) return "null";

    const t = typeof v;
    if (t === "string") return v;
    if (t === "number" || t === "boolean" || t === "bigint") return String(v);

    if (isThrowableLike(v)) return throwableToText(v);

    if (t === "object") return safeJson(v);

    try {
        return String(v);
    } catch (_) {
    }
    return "[unserializable]";
}

function joinArgs(args, from, toExclusive) {
    let out = "";
    for (let i = from; i < toExclusive; i++) {
        out += (i > from ? " " : "") + valueToText(args[i]);
    }
    return out;
}

function makePrefix(scope) {
    const s = String(scope || "").trim();
    return s ? "[" + s + "] " : "";
}

function makeApi(engine /*, K */) {
    const log = (engine && engine.log && typeof engine.log === "function") ? engine.log() : null;

    function has(fn) {
        return !!(log && typeof log[fn] === "function");
    }

    function call1(levelFn, msg) {
        if (!log) return;
        if (has(levelFn)) log[levelFn](msg);
        else if (has("info")) log.info(msg);
    }

    /**
     * Important: we NEVER call log[levelFn](msg, err) here.
     * Java side may expose only 1-arity overload and Graal interop may throw uncatchable arity errors.
     */
    function write(levelFn, scope, args) {
        const prefix = makePrefix(scope);

        if (!args || args.length === 0) {
            const msg0 = prefix;
            try {
                call1(levelFn, msg0);
            } catch (_) {
            }
            return msg0;
        }

        // If last arg is throwable-like, append it on a new line.
        if (args.length >= 2 && isThrowableLike(args[args.length - 1])) {
            const head = prefix + joinArgs(args, 0, args.length - 1);
            const err = args[args.length - 1];
            const msg = head + "\n" + throwableToText(err);

            try {
                call1(levelFn, msg);
            } catch (_) {
            }
            return msg;
        }

        const msg = prefix + joinArgs(args, 0, args.length);
        try {
            call1(levelFn, msg);
        } catch (_) {
        }
        return msg;
    }

    function trace() {
        return write("trace", "", arguments);
    }

    function debug() {
        return write("debug", "", arguments);
    }

    function info() {
        return write("info", "", arguments);
    }

    function warn() {
        return write("warn", "", arguments);
    }

    function error() {
        return write("error", "", arguments);
    }

    function fatal() {
        return write("fatal", "", arguments);
    }

    function scoped(scopeName) {
        const scope = String(scopeName || "").trim();
        return Object.freeze({
            trace: function () {
                return write("trace", scope, arguments);
            },
            debug: function () {
                return write("debug", scope, arguments);
            },
            info: function () {
                return write("info", scope, arguments);
            },
            warn: function () {
                return write("warn", scope, arguments);
            },
            error: function () {
                return write("error", scope, arguments);
            },
            fatal: function () {
                return write("fatal", scope, arguments);
            },
            scope: scope
        });
    }

    function enabled() {
        return !!log;
    }

    return Object.freeze({
        enabled: enabled,
        trace: trace,
        debug: debug,
        info: info,
        warn: warn,
        error: error,
        fatal: fatal,
        child: scoped,
        scope: scoped,
        safeJson: safeJson
    });
}

function create(engine, K) {
    if (!engine) throw new Error("[LOG] engine is required");
    return makeApi(engine, K);
}

create.META = {
    moduleId: "log",
    globalName: "LOG",
    version: "1.2.0",
    description: "Rootkit wrapper for engine.log() with safe formatting + scoped child loggers",
    engineMin: "0.1.0"
};

module.exports = create;

File: kalitech-engine/src/main/resources/kalitech/engine/modules/Log/Log.lua
local M = {}
local json = require("@builtin/json")
local luaRuntime = require("@builtin/lua_runtime")
local Error = luaRuntime.Error
local LuaInstanceOf = luaRuntime.LuaInstanceOf
local LuaTypeOf = luaRuntime.LuaTypeOf
local LuaStringTrim = luaRuntime.LuaStringTrim
local LuaConstruct = luaRuntime.LuaConstruct
function safeJson(self, v)
    do
        local lua_try, lua_hasReturned, lua_returnValue = pcall(function()
            return true, json:encode(v)
        end)
        if lua_try and lua_hasReturned then
            return lua_returnValue
        end
    end
    do
        local lua_try, lua_hasReturned, lua_returnValue = pcall(function()
            return true, tostring(v)
        end)
        if lua_try and lua_hasReturned then
            return lua_returnValue
        end
    end
    return "[unserializable]"
end
function isThrowableLike(self, v)
    if not v then
        return false
    end
    do
        local lua_try, lua_hasReturned, lua_returnValue = pcall(function()
            if LuaInstanceOf(v, Error) then
                return true, true
            end
        end)
        if lua_try and lua_hasReturned then
            return lua_returnValue
        end
    end
    local t = LuaTypeOf(v)
    if t ~= "object" and t ~= "function" then
        return false
    end
    do
        local lua_try, lua_hasReturned, lua_returnValue = pcall(function()
            local stack = v.stack
            if KTypeOf(stack) == "string" and #stack > 0 then
                return true, true
            end
        end)
        if lua_try and lua_hasReturned then
            return lua_returnValue
        end
    end
    do
        local lua_try, lua_hasReturned, lua_returnValue = pcall(function()
            local name = v.name
            local msg = v.message
            if KTypeOf(name) == "string" and #name > 0 and KTypeOf(msg) == "string" then
                return true, true
            end
        end)
        if lua_try and lua_hasReturned then
            return lua_returnValue
        end
    end
    do
        local lua_try, lua_hasReturned, lua_returnValue = pcall(function()
            if KTypeOf(v.getClass) == "function" then
                return true, true
            end
        end)
        if lua_try and lua_hasReturned then
            return lua_returnValue
        end
    end
    return false
end
function throwableToText(self, e)
    if not e then
        return "null"
    end
    do
        local lua_try, lua_hasReturned, lua_returnValue = pcall(function()
            if LuaInstanceOf(e, Error) then
                return true, e.stack or e.message or tostring(e)
            end
        end)
        if lua_try and lua_hasReturned then
            return lua_returnValue
        end
    end
    do
        local lua_try, lua_hasReturned, lua_returnValue = pcall(function()
            if KTypeOf(e.stack) == "string" and e.stack then
                return true, e.stack
            end
        end)
        if lua_try and lua_hasReturned then
            return lua_returnValue
        end
    end
    do
        local lua_try, lua_hasReturned, lua_returnValue = pcall(function()
            if KTypeOf(e.message) == "string" and e.message then
                return true, e.message
            end
        end)
        if lua_try and lua_hasReturned then
            return lua_returnValue
        end
    end
    do
        local lua_try, lua_hasReturned, lua_returnValue = pcall(function()
            return true, tostring(e)
        end)
        if lua_try and lua_hasReturned then
            return lua_returnValue
        end
    end
    return "[unserializable-exception]"
end
function valueToText(self, v)
    if v == nil then
        return "null"
    end
    local t = LuaTypeOf(v)
    if t == "string" then
        return v
    end
    if t == "number" or t == "boolean" or t == "bigint" then
        return tostring(v)
    end
    if isThrowableLike(_G, v) then
        return throwableToText(_G, v)
    end
    if t == "object" then
        return safeJson(_G, v)
    end
    do
        local lua_try, lua_hasReturned, lua_returnValue = pcall(function()
            return true, tostring(v)
        end)
        if lua_try and lua_hasReturned then
            return lua_returnValue
        end
    end
    return "[unserializable]"
end
function joinArgs(self, args, from, toExclusive)
    local out = ""
    do
        local i = from
        while i < toExclusive do
            out = out .. (i > from and " " or "") .. tostring(valueToText(_G, KIndex(args, i)))
            i = i + 1
        end
    end
    return out
end
function makePrefix(self, scope)
    local s = LuaStringTrim(tostring(scope or ""))
    local lua_s_0
    if s then
        lua_s_0 = ("[" .. s) .. "] "
    else
        lua_s_0 = ""
    end
    return lua_s_0
end
function makeApi(self, engine)
    local lua_temp_1
    if engine and engine.log and KTypeOf(engine.log) == "function" then
        lua_temp_1 = engine:log()
    else
        lua_temp_1 = nil
    end
    local log = lua_temp_1
    local function has(self, fn)
        return not not (log and KTypeOf(log[fn]) == "function")
    end
    local function call1(self, levelFn, msg)
        if not log then
            return
        end
        if has(_G, levelFn) then
            log[levelFn](log, msg)
        elseif has(_G, "info") then
            log:info(msg)
        end
    end
    local function write(self, levelFn, scope, args)
        local prefix = makePrefix(_G, scope)
        if not args or KLength(args) == 0 then
            local msg0 = prefix
            do
                pcall(function()
                    call1(_G, levelFn, msg0)
                end)
            end
            return msg0
        end
        if KLength(args) >= 2 and isThrowableLike(_G, KIndex(args, KLength(args) - 1)) then
            local head = prefix .. joinArgs(_G, args, 0, KLength(args) - 1)
            local err = KIndex(args, KLength(args) - 1)
            local msg = (head .. "\n") .. tostring(throwableToText(_G, err))
            do
                pcall(function()
                    call1(_G, levelFn, msg)
                end)
            end
            return msg
        end
        local msg = prefix .. joinArgs(_G, args, 0, KLength(args))
        do
            pcall(function()
                call1(_G, levelFn, msg)
            end)
        end
        return msg
    end
    local function trace(self, ...)
        local args = {...}
        return write(_G, "trace", "", args)
    end
    local function lua_debug(self, ...)
        local args = {...}
        return write(_G, "debug", "", args)
    end
    local function info(self, ...)
        local args = {...}
        return write(_G, "info", "", args)
    end
    local function warn(self, ...)
        local args = {...}
        return write(_G, "warn", "", args)
    end
    local function lua_error(self, ...)
        local args = {...}
        return write(_G, "error", "", args)
    end
    local function fatal(self, ...)
        local args = {...}
        return write(_G, "fatal", "", args)
    end
    local function scoped(self, scopeName)
        local scope = LuaStringTrim(tostring(scopeName or ""))
        return KObject:freeze({
            trace = function(self, ...)
                local args = {...}
                return write(_G, "trace", scope, args)
            end,
            debug = function(self, ...)
                local args = {...}
                return write(_G, "debug", scope, args)
            end,
            info = function(self, ...)
                local args = {...}
                return write(_G, "info", scope, args)
            end,
            warn = function(self, ...)
                local args = {...}
                return write(_G, "warn", scope, args)
            end,
            error = function(self, ...)
                local args = {...}
                return write(_G, "error", scope, args)
            end,
            fatal = function(self, ...)
                local args = {...}
                return write(_G, "fatal", scope, args)
            end,
            scope = scope
        })
    end
    local function enabled(self)
        return not not log
    end
    return KObject:freeze({
        enabled = enabled,
        trace = trace,
        debug = lua_debug,
        info = info,
        warn = warn,
        error = lua_error,
        fatal = fatal,
        child = scoped,
        scope = scoped,
        safeJson = safeJson
    })
end
create = setmetatable(
    {},
    {__call = function(lua_, self, engine, K)
        if not engine then
            error(
                LuaConstruct(Error, "[LOG] engine is required"),
                0
            )
        end
        return makeApi(_G, engine, K)
    end}
)
create.META = {
    moduleId = "log",
    version = "1.2.0",
    description = "Rootkit wrapper for engine.log() with safe formatting + scoped child loggers",
    engineMin = "0.1.0"
}
M = create

return M
